README.md
# maxminddb

Pure Ruby [GeoIP2 MaxMind DB](http://maxmind.github.io/MaxMind-DB/) reader, which doesn't require [libmaxminddb](https://github.com/maxmind/libmaxminddb).

You can find more information about the GeoIP2 database [here](http://dev.maxmind.com/geoip/geoip2/downloadable/).

[![Gem Version](https://badge.fury.io/rb/maxminddb.svg)](http://badge.fury.io/rb/maxminddb)
[![Build Status](https://travis-ci.org/yhirose/maxminddb.svg?branch=master)](https://travis-ci.org/yhirose/maxminddb)
[![Code Climate](https://codeclimate.com/github/yhirose/maxminddb.png)](https://codeclimate.com/github/yhirose/maxminddb)

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'maxminddb'
```

And then execute:

```sh
$ bundle
```

Or install it yourself as:

```sh
$ gem install maxminddb
```

## Usage

```ruby
db = MaxMindDB.new('./GeoLite2-City.mmdb')
ret = db.lookup('74.125.225.224')

ret.found? # => true
ret.country.name # => 'United States'
ret.country.name('zh-CN') # => '美国'
ret.country.iso_code # => 'US'
ret.city.name(:fr) # => 'Mountain View'
ret.subdivisions.most_specific.name # => 'California'
ret.location.latitude # => -122.0574
ret.to_hash # => {"city"=>{"geoname_id"=>5375480, "names"=>{"de"=>"Mountain View", "en"=>"Mountain View", "fr"=>"Mountain View", "ja"=>"マウンテンビュー", "ru"=>"Маунтин-Вью", "zh-CN"=>"芒廷维尤"}}, "continent"=>{"code"=>"NA", "geoname_id"=>6255149, "names"=>{"de"=>"Nordamerika", "en"=>"North America", "es"=>"Norteamérica", "fr"=>"Amérique du Nord", "ja"=>"北アメリカ", "pt-BR"=>"América do Norte", "ru"=>"Северная Америка", "zh-CN"=>"北美洲"}}, "country"=>{"geoname_id"=>6252001, "iso_code"=>"US", "names"=>{"de"=>"USA", "en"=>"United States", "es"=>"Estados Unidos", "fr"=>"États-Unis", "ja"=>"アメリカ合衆国", "pt-BR"=>"Estados Unidos", "ru"=>"Сша", "zh-CN"=>"美国"}}, "location"=>{"latitude"=>37.419200000000004, "longitude"=>-122.0574, "metro_code"=>807, "time_zone"=>"America/Los_Angeles"}, "postal"=>{"code"=>"94043"}, "registered_country"=>{"geoname_id"=>6252001, "iso_code"=>"US", "names"=>{"de"=>"USA", "en"=>"United States", "es"=>"Estados Unidos", "fr"=>"États-Unis", "ja"=>"アメリカ合衆国", "pt-BR"=>"Estados Unidos", "ru"=>"Сша", "zh-CN"=>"美国"}}, "subdivisions"=>[{"geoname_id"=>5332921, "iso_code"=>"CA", "names"=>{"de"=>"Kalifornien", "en"=>"California", "es"=>"California", "fr"=>"Californie", "ja"=>"カリフォルニア州", "pt-BR"=>"Califórnia", "ru"=>"Калифорния", "zh-CN"=>"加利福尼亚州"}}]}
```

Even if no result could be found, you can ask for the attributes without guarding for nil:

```ruby
db = MaxMindDB.new('./GeoLite2-City.mmdb')
ret = db.lookup('127.0.0.1')
ret.found? # => false
ret.country.name # => nil
ret.to_hash # => {}
```

For testing or other purposes, you might wish to treat localhost IP addresses as some other address - an external one. You can do this by assigning the desired external IP address to the attribute local_ip_alias:

```ruby
db = MaxMindDB.new('./GeoLite2-City.mmdb')
ret = db.local_ip_alias = '74.125.225.224'
ret = db.lookup('127.0.0.1')
ret.found? # => true
ret.country.name # => 'United States'
ret.to_hash.empty? # => false
```

It's also possible to access the database metadata.

```ruby
db = MaxMindDB.new('./GeoLite2-City.mmdb')
db.metadata['build_epoch'] # => 1493762948
db.metadata # => {"binary_format_major_version"=>2, "binary_format_minor_version"=>0, "build_epoch"=>1493762948, "database_type"=>"GeoLite2-City", "description"=>{"en"=>"GeoLite2 City database"}, "ip_version"=>6, "languages"=>["de", "en", "es", "fr", "ja", "pt-BR", "ru", "zh-CN"], "node_count"=>3678850, "record_size"=>28}
```

### Regarding thread safety

A MaxMindDB instance doesn't do any write operation after it is created. So we can consider it as an immutable object which is 'thread-safe'.

### JSON web server on Docker

maxminddb-docker: https://github.com/samnissen/maxminddb-docker

### File reading strategies

By default, `MaxMinDB.new` will read the entire database into memory. This makes subsequent lookups fast, but can result in a fairly large memory overhead.

If having a low memory overhead is important, you can use the `LowMemoryReader` by passing a `file_reader` argument to `MaxMindDB.new`. For example:

```ruby
db = MaxMindDB.new('./GeoLite2-City.mmdb', MaxMindDB::LOW_MEMORY_FILE_READER)
ret = db.lookup('74.125.225.224')
```

The `LowMemoryReader` will not load the entire database into memory. It's important to note that for Ruby versions lower than `2.5.0`, the `LowMemoryReader` is not process safe. Forking a process after initializing a `MaxMindDB` instance can lead to unexpected results. For Ruby versions >= `2.5.0`, `LowMemoryReader` uses `File.pread` which works safely in forking environments.

## Contributing

1.  Fork it ( http://github.com/yhirose/maxminddb/fork )
2.  Create your feature branch (`git checkout -b my-new-feature`)
3.  Commit your changes (`git commit -am 'Add some feature'`)
4.  Push to the branch (`git push origin my-new-feature`)
5.  Create new Pull Request


CHANGELOG.md
# Changelog

### 0.1.21 (July 20, 2018)

- Make the low-memory file reader opt-in

### 0.1.20 (July 9, 2018)

- Add local_ip_alias

### 0.1.19 (June 22, 2018)

- Implement a faster default reader

### 0.1.18 (June 18, 2018)

- Backward Compatible Client Initialize

### 0.1.17 (June 9, 2018)

- Allow the database file reader to be overridden.
- increase coverage for IPv6 and EU tests

### 0.1.16 (May 22, 2018)

- Adds is_in_european_union feature

### 0.1.15 (December 18, 2017)

- Implement #most_specific method for result subdivisions #31

### 0.1.14 (July 26, 2017)

- Expose accuracy_radius in Location data #29

### 0.1.13 (June 13, 2017)

- Expose database metadata #27

### 0.1.12 (January 11, 2017)

- Removed Fixnum and Bignum deprecation warnings (in ruby 2.4)
- Update Travis
- Fix failing test

### 0.1.11 (March 7, 2016)

- Fixed problem with 32bit record data

### 0.1.10 (February 25, 2016)

- Performance tweaks

***

*For older releases, see commit log.*


lib/maxminddb.rb
require "maxminddb/version"
require 'maxminddb/result'
require 'maxminddb/reader'
require 'ipaddr'

module MaxMindDB

  # The default reader for MaxMindDB files. Reads the database into memory.
  # This creates a higher memory overhead, but faster lookup times.
  DEFAULT_FILE_READER = proc { |path| File.binread(path) }

  # A low memory file reader for MaxMindDB files. Avoids reading the database
  # into memory. Has a lower memory footprint but slower lookup times.
  LOW_MEMORY_FILE_READER = proc { |path| MaxMindDB::LowMemoryReader.new(path) }

  def self.new(path, file_reader=DEFAULT_FILE_READER)
    Client.new(path, file_reader)
  end

  class Client
    METADATA_BEGIN_MARKER = ([0xAB, 0xCD, 0xEF].pack('C*') + 'MaxMind.com').encode('ascii-8bit', 'ascii-8bit')
    DATA_SECTION_SEPARATOR_SIZE = 16
    SIZE_BASE_VALUES = [0, 29, 285, 65821]
    POINTER_BASE_VALUES = [0, 0, 2048, 526336]

    attr_reader :metadata

    # An IP that is used instead of local IPs
    attr_accessor :local_ip_alias

    def initialize(path, file_reader = DEFAULT_FILE_READER)
      @path = path
      @data = file_reader.call(path)

      pos = @data.rindex(METADATA_BEGIN_MARKER)
      raise 'invalid file format' unless pos
      pos += METADATA_BEGIN_MARKER.size
      @metadata = decode(pos, 0)[1]

      @ip_version = @metadata['ip_version']
      @node_count = @metadata['node_count']
      @node_byte_size = @metadata['record_size'] * 2 / 8
      @search_tree_size = @node_count * @node_byte_size
    end

    def inspect
      "#<MaxMindDB::Client: DBPath:'#{@path}'>"
    end

    def lookup(ip_or_hostname)
      if @local_ip_alias && is_local?(ip_or_hostname)
        ip_or_hostname = @local_ip_alias
      end

      node_no = 0
      addr = addr_from_ip(ip_or_hostname)
      start_idx = @ip_version == 4 ? 96 : 0
      for i in start_idx ... 128
        flag = (addr >> (127 - i)) & 1
        next_node_no = read_record(node_no, flag)
        if next_node_no == 0
          raise 'invalid file format'
        elsif next_node_no >= @node_count
          data_section_start = @search_tree_size + DATA_SECTION_SEPARATOR_SIZE
          pos = (next_node_no - @node_count) - DATA_SECTION_SEPARATOR_SIZE
          return MaxMindDB::Result.new(decode(pos, data_section_start)[1])
        else
          node_no = next_node_no
        end
      end
      raise 'invalid file format'
    end

    private

    def read_record(node_no, flag)
      rec_byte_size = @node_byte_size / 2
      pos = @node_byte_size * node_no
      middle = @data[pos + rec_byte_size].ord if @node_byte_size.odd?
      if flag == 0 # left
        val = read_value(pos, 0, rec_byte_size)
        val += ((middle & 0xf0) << 20) if middle
      else # right
        val = read_value(pos + @node_byte_size - rec_byte_size, 0, rec_byte_size)
        val += ((middle & 0xf) << 24) if middle
      end
      val
    end

    def decode(pos, base_pos)
      ctrl = @data[pos + base_pos].ord
      pos += 1

      type = ctrl >> 5

      if type == 1 # pointer
        size = ((ctrl >> 3) & 0x3) + 1
        v1 = ctrl & 0x7
        v2 = read_value(pos, base_pos, size)
        pos += size

        pointer = (v1 << (8 * size)) + v2 + POINTER_BASE_VALUES[size]
        val = decode(pointer, base_pos)[1]
      else
        if type == 0 # extended type
          type = 7 + @data[pos + base_pos].ord
          pos += 1
        end

        size = ctrl & 0x1f
        if size >= 29
          byte_size = size - 29 + 1
          val = read_value(pos, base_pos, byte_size)
          pos += byte_size
          size = val + SIZE_BASE_VALUES[byte_size]
        end

        case type
        when 2 # utf8
          val = @data[pos + base_pos, size].encode('utf-8', 'utf-8')
          pos += size
        when 3 # double
          val = @data[pos + base_pos, size].unpack('G')[0]
          pos += size
        when 4 # bytes
          val = @data[pos + base_pos, size]
          pos += size
        when 5 # unsigned 16-bit int
          val = read_value(pos, base_pos, size)
          pos += size
        when 6 # unsigned 32-bit int
          val = read_value(pos, base_pos, size)
          pos += size
        when 7 # map
          val = {}
          size.times do
            pos, k = decode(pos, base_pos)
            pos, v = decode(pos, base_pos)
            val[k] = v
          end
        when 8 # signed 32-bit int
          v1 = @data[pos + base_pos, size].unpack('N')[0]
          bits = size * 8
          val = (v1 & ~(1 << bits)) - (v1 & (1 << bits))
          pos += size
        when 9 # unsigned 64-bit int
          val = read_value(pos, base_pos, size)
          pos += size
        when 10 # unsigned 128-bit int
          val = read_value(pos, base_pos, size)
          pos += size
        when 11 # array
          val = []
          size.times do
            pos, v = decode(pos, base_pos)
            val.push(v)
          end
        when 12 # data cache container
          raise 'TODO:'
        when 13 # end marker
          val = nil
        when 14 # boolean
          val = (size != 0)
        when 15 # float
          val = @data[pos + base_pos, size].unpack('g')[0]
          pos += size
        end
      end

      [pos, val]
    end

    def read_value(pos, base_pos, size)
      bytes = @data[pos + base_pos, size].unpack('C*')
      bytes.inject(0){|r, v| (r << 8) + v }
    end

    def addr_from_ip(ip_or_hostname)
      klass = ip_or_hostname.class

      return ip_or_hostname if RUBY_VERSION.to_f < 2.4 && (klass == Fixnum || klass == Bignum)
      return ip_or_hostname if RUBY_VERSION.to_f >= 2.4 && klass == Integer

      addr = IPAddr.new(ip_or_hostname)
      addr = addr.ipv4_compat if addr.ipv4?
      addr.to_i
    end

    def is_local?(ip_or_hostname)
      ["127.0.0.1", "localhost", "::1", "0000::1", "0:0:0:0:0:0:0:1"].include? ip_or_hostname
    end
  end
end

# vim: et ts=2 sw=2 ff=unix
